data race in TestApplySnapshotAndCommittedEntries

```
==================
WARNING: DATA RACE
Write by goroutine 12:
  github.com/coreos/etcd/etcdserver.(*raftNode).run()
      github.com/coreos/etcd/etcdserver/_test/_obj_test/raft.go:90 +0x9d
Previous read by goroutine 11:
  github.com/coreos/etcd/etcdserver.func·003()
      github.com/coreos/etcd/etcdserver/_test/_obj_test/server.go:472 +0x7e
  github.com/coreos/etcd/etcdserver.(*EtcdServer).run()
      github.com/coreos/etcd/etcdserver/_test/_obj_test/server.go:556 +0x171d
Goroutine 12 (running) created at:
  github.com/coreos/etcd/etcdserver.(*EtcdServer).run()
      github.com/coreos/etcd/etcdserver/_test/_obj_test/server.go:469 +0x338
Goroutine 11 (running) created at:
  github.com/coreos/etcd/etcdserver.(*EtcdServer).start()
      github.com/coreos/etcd/etcdserver/_test/_obj_test/server.go:386 +0x58f
  github.com/coreos/etcd/etcdserver.TestApplySnapshotAndCommittedEntries()
      /home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server_test.go:821 +0x614
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:447 +0x133
==================
panic: test timed out after 3m0s
goroutine 75 [running]:
testing.func·008()
	/usr/local/go/src/testing/testing.go:681 +0x168
created by time.goFunc
	/usr/local/go/src/time/sleep.go:129 +0x59
goroutine 1 [chan receive]:
testing.RunTests(0xaebf78, 0x16020c0, 0x35, 0x35, 0xc208068201)
	/usr/local/go/src/testing/testing.go:556 +0xdc3
testing.(*M).Run(0xc20806a320, 0x160cc60)
	/usr/local/go/src/testing/testing.go:485 +0xe1
main.main()
	github.com/coreos/etcd/etcdserver/_test/_testmain.go:218 +0x382
goroutine 86 [chan send (nil chan)]:
github.com/coreos/etcd/etcdserver.func·003()
	github.com/coreos/etcd/etcdserver/_test/_obj_test/server.go:472 +0xa9
github.com/coreos/etcd/etcdserver.(*EtcdServer).run(0xc208167440)
	github.com/coreos/etcd/etcdserver/_test/_obj_test/server.go:556 +0x171e
created by github.com/coreos/etcd/etcdserver.(*EtcdServer).start
	github.com/coreos/etcd/etcdserver/_test/_obj_test/server.go:386 +0x58f
goroutine 85 [chan receive]:
github.com/coreos/etcd/etcdserver.(*EtcdServer).Stop(0xc208167440)
	github.com/coreos/etcd/etcdserver/_test/_obj_test/server.go:571 +0x188
github.com/coreos/etcd/etcdserver.TestApplySnapshotAndCommittedEntries(0xc208138900)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server_test.go:831 +0x85c
testing.tRunner(0xc208138900, 0x16024c8)
	/usr/local/go/src/testing/testing.go:447 +0x134
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:555 +0xd4e
goroutine 87 [select]:
github.com/coreos/etcd/etcdserver.(*raftNode).run(0xc208167450)
	github.com/coreos/etcd/etcdserver/_test/_obj_test/raft.go:128 +0xd9b
created by github.com/coreos/etcd/etcdserver.(*EtcdServer).run
	github.com/coreos/etcd/etcdserver/_test/_obj_test/server.go:469 +0x338
FAIL	github.com/coreos/etcd/etcdserver	180.023s
```

--- a/etcdserver/server_test.go
+++ b/etcdserver/server_test.go
@@ -607,7 +607,7 @@ func TestSync(t *testing.T) {
 	})
 	srv.sync(10 * time.Second)
 	timer.Stop()
-	testutil.ForceGosched()
+	testutil.WaitSchedule()
 
 	action := n.Action()
 	if len(action) != 1 {
@@ -642,7 +642,7 @@ func TestSyncTimeout(t *testing.T) {
 	timer.Stop()
 
 	// give time for goroutine in sync to cancel
-	testutil.ForceGosched()
+	testutil.WaitSchedule()
 	w := []testutil.Action{{Name: "Propose blocked"}}
 	if g := n.Action(); !reflect.DeepEqual(g, w) {
 		t.Errorf("action = %v, want %v", g, w)
@@ -676,7 +676,7 @@ func TestSyncTrigger(t *testing.T) {
 	}
 	// trigger a sync request
 	st <- time.Time{}
-	testutil.ForceGosched()
+	testutil.WaitSchedule()
 
 	action := n.Action()
 	if len(action) != 1 {
@@ -710,7 +710,7 @@ func TestSnapshot(t *testing.T) {
 		store: st,
 	}
 	srv.snapshot(1, raftpb.ConfState{Nodes: []uint64{1}})
-	testutil.ForceGosched()
+	testutil.WaitSchedule()
 	gaction := st.Action()
 	if len(gaction) != 2 {
 		t.Fatalf("len(action) = %d, want 1", len(gaction))
@@ -786,7 +786,7 @@ func TestRecvSnapshot(t *testing.T) {
 	s.start()
 	n.readyc <- raft.Ready{Snapshot: raftpb.Snapshot{Metadata: raftpb.SnapshotMetadata{Index: 1}}}
 	// make goroutines move forward to receive snapshot
-	testutil.ForceGosched()
+	testutil.WaitSchedule()
 	s.Stop()
 
 	wactions := []testutil.Action{{Name: "Recovery"}}
@@ -827,7 +827,7 @@ func TestApplySnapshotAndCommittedEntries(t *testing.T) {
 		},
 	}
 	// make goroutines move forward to receive snapshot
-	testutil.ForceGosched()
+	testutil.WaitSchedule()
 	s.Stop()
 
 	actions := st.Action()

--- a/pkg/testutil/testutil.go
+++ b/pkg/testutil/testutil.go
@@ -16,17 +16,13 @@ package testutil
 
 import (
 	"net/url"
-	"runtime"
 	"testing"
+	"time"
 )
 
-// WARNING: This is a hack.
-// Remove this when we are able to block/check the status of the go-routines.
-func ForceGosched() {
-	// possibility enough to sched up to 10 go routines.
-	for i := 0; i < 10000; i++ {
-		runtime.Gosched()
-	}
+// TODO: improve this when we are able to know the schedule or status of target go-routine.
+func WaitSchedule() {
+	time.Sleep(3 * time.Millisecond)
 }
 
 func MustNewURLs(t *testing.T, urls []string) []url.URL {

--- a/raft/node_test.go
+++ b/raft/node_test.go
@@ -199,7 +199,7 @@ func TestBlockProposal(t *testing.T) {
 		errc <- n.Propose(context.TODO(), []byte("somedata"))
 	}()
 
-	testutil.ForceGosched()
+	testutil.WaitSchedule()
 	select {
 	case err := <-errc:
 		t.Errorf("err = %v, want blocking", err)
@@ -207,7 +207,7 @@ func TestBlockProposal(t *testing.T) {
 	}
 
 	n.Campaign(context.TODO())
-	testutil.ForceGosched()
+	testutil.WaitSchedule()
 	select {
 	case err := <-errc:
 		if err != nil {

--- a/rafthttp/pipeline_test.go
+++ b/rafthttp/pipeline_test.go
@@ -39,7 +39,7 @@ func TestPipelineSend(t *testing.T) {
 	p := newPipeline(tr, picker, types.ID(2), types.ID(1), types.ID(1), fs, &fakeRaft{}, nil)
 
 	p.msgc <- raftpb.Message{Type: raftpb.MsgApp}
-	testutil.ForceGosched()
+	testutil.WaitSchedule()
 	p.stop()
 
 	if tr.Request() == nil {
@@ -60,7 +60,7 @@ func TestPipelineExceedMaximalServing(t *testing.T) {
 
 	// keep the sender busy and make the buffer full
 	// nothing can go out as we block the sender
-	testutil.ForceGosched()
+	testutil.WaitSchedule()
 	for i := 0; i < connPerPipeline+pipelineBufSize; i++ {
 		select {
 		case p.msgc <- raftpb.Message{}:
@@ -68,7 +68,7 @@ func TestPipelineExceedMaximalServing(t *testing.T) {
 			t.Errorf("failed to send out message")
 		}
 		// force the sender to grab data
-		testutil.ForceGosched()
+		testutil.WaitSchedule()
 	}
 
 	// try to send a data when we are sure the buffer is full
@@ -80,7 +80,7 @@ func TestPipelineExceedMaximalServing(t *testing.T) {
 
 	// unblock the senders and force them to send out the data
 	tr.unblock()
-	testutil.ForceGosched()
+	testutil.WaitSchedule()
 
 	// It could send new data after previous ones succeed
 	select {
@@ -99,7 +99,7 @@ func TestPipelineSendFailed(t *testing.T) {
 	p := newPipeline(newRespRoundTripper(0, errors.New("blah")), picker, types.ID(2), types.ID(1), types.ID(1), fs, &fakeRaft{}, nil)
 
 	p.msgc <- raftpb.Message{Type: raftpb.MsgApp}
-	testutil.ForceGosched()
+	testutil.WaitSchedule()
 	p.stop()
 
 	fs.Lock()

--- a/rafthttp/stream_test.go
+++ b/rafthttp/stream_test.go
@@ -33,7 +33,7 @@ func TestStreamWriterAttachOutgoingConn(t *testing.T) {
 		prevwfc := wfc
 		wfc = &fakeWriteFlushCloser{}
 		sw.attach(&outgoingConn{t: streamTypeMessage, Writer: wfc, Flusher: wfc, Closer: wfc})
-		testutil.ForceGosched()
+		testutil.WaitSchedule()
 		// previous attached connection should be closed
 		if prevwfc != nil && prevwfc.closed != true {
 			t.Errorf("#%d: close of previous connection = %v, want true", i, prevwfc.closed)
@@ -44,7 +44,7 @@ func TestStreamWriterAttachOutgoingConn(t *testing.T) {
 		}
 
 		sw.msgc <- raftpb.Message{}
-		testutil.ForceGosched()
+		testutil.WaitSchedule()
 		// still working
 		if _, ok := sw.writec(); ok != true {
 			t.Errorf("#%d: working status = %v, want true", i, ok)
@@ -73,7 +73,7 @@ func TestStreamWriterAttachBadOutgoingConn(t *testing.T) {
 	sw.attach(&outgoingConn{t: streamTypeMessage, Writer: wfc, Flusher: wfc, Closer: wfc})
 
 	sw.msgc <- raftpb.Message{}
-	testutil.ForceGosched()
+	testutil.WaitSchedule()
 	// no longer working
 	if _, ok := sw.writec(); ok != false {
 		t.Errorf("working = %v, want false", ok)

--- a/rafthttp/transport_test.go
+++ b/rafthttp/transport_test.go
@@ -137,7 +137,7 @@ func TestTransportErrorc(t *testing.T) {
 	}
 	tr.peers[1].Send(raftpb.Message{})
 
-	testutil.ForceGosched()
+	testutil.WaitSchedule()
 	select {
 	case <-errorc:
 	default:

